### ISSUE TITLE
Question on baseline system (LSTM or GRU model)

### ISSUE THREAD

**Issue by kelvinqin** [simonalexanderson/StyleGestures#24]
Hi Smion,
I am trying to build a baseline system (GRU model) following your paper, my initial implementation generated pose which is very dynamic (arms move quickly in large scale in my upper body model; body turn around quickly in a crazy manner in my full-body model). I am guessing maybe it is because I have over-trained the model. So I started to look into your experiments detail in your paper (table below)

![image](https://user-images.githubusercontent.com/10486482/104370181-721ebf00-5559-11eb-8296-dc6ff962023f.png)

In my case, I am using uni-direction GRU with 2 layer and 512 hidden_size, my batch_size is 80 and num_batches is 80000, and in terms of # of epoches, it is 1213 epoches, but seems you only run 50 epoches, I am wondering what is the reason for you to only run 50 epoches for your LSTM baseline? How many batch-steps does "50 epoches" equal to in your case? Did you also experience over-train as I did if you run more epoches?

I will test your setting also (unidirection LSTM with 1 layer and 350 hidden size) and run less epoches (say 50) and let you know the result. But I am curious the logic of your setting. Appreciated.

Thanks!
Kelvin

**Comment by ghenter** [comment]
> I am guessing maybe it is because I have over-trained the model.

By "over-training", do you mean "overfitting", or something else? The term "over-training" is not standard in the literature I have been reading.

If you suspect overfitting, I would recommend that you generate output motion at regular intervals during training and see how the quality changes. If the motion quality first improves but then starts getting worse again, then you have some form of overfitting, otherwise you do not. But if the quality never is good enough, then your issues cannot be attributed to overfitting alone.

Example motion from our trained system should be available in our supplementary material, available in the [Computer Graphics Forum digital repository](https://diglib.eg.org/handle/10.1111/cgf13946). I personally don't know the exact motivation for the specific training duration etc. that Simon chose. However, I think that will be more efficient and instructive for you to run your own experiments (e.g., as suggested in the previous paragraph) than to wait for Simon to weigh in with his perspective.

**Comment by kelvinqin** [comment]
Hi, Gustav,
Thanks for your answering, yes, what I mean is "overfitting", thanks!

Yes, I will wait for more information from Simon about why he run so few number of epoches for the baseline system. (LSTM)

BTW, what is the exact LSTM/GRU Simon is using for that baseline experiment? I can find two classes defined (LSTM and GRU) in glow/modules.py, not sure if that is the exact model he is using. 

Have a nice day,

**Comment by ghenter** [comment]
> what is the exact LSTM/GRU Simon is using for that baseline experiment? I can find two classes defined (LSTM and GRU) in glow/modules.py, not sure if that is the exact model he is using.

The LSTM baseline is a very simple network (a single-layer, 350 unit unidirectional LSTM, as described in Sec. 4.4 of our paper), without any autoregression or normalising flows. At least in the original MoGlow paper, the RNN (LSTM) baseline system used a completely different codebase than the MG systems, and I would suspect the same is true here. I did a quick search for "LSTM" in `glow/modules.py`, and nothing in there seems relevant; I only spot references to using LSTMs (or GRUs) _within the affine coupling in MoGlow_, which is not relevant to the baseline since the baseline doesn't use normalising flows at all!

Because the LSTM baseline from the paper is trivial to implement in something like Keras, I don't think it is included in this repository.

> I will wait for more information from Simon about why he run so few number of epoches for the baseline system. (LSTM)

Like I emphasised in my previous post, you can test your hypothesis regarding overfitting without waiting for Simon to respond.

My guess is that he ran so few epochs because the simplistic baseline system – which has just a single layer, no autoregression, and no normalising flows at all – converges much faster than the MoGlow model.

**Comment by simonalexanderson** [comment]
I can confirm Gustav's comment; the LSTM system is really a bottom baseline and converges fast. It proves the point of MSE loss collapsing to the mean pose. There is no code for it in the repository.
